### PR TITLE
Interop WIP

### PR DESCRIPTION
**Pull request by mpetrunic**
initial setup
```
cd lodestar/packages/lodestar
yarn global add lerna
lerna bootstrap
```
run lodestar
```
# necessary options
# -p minimal
# -d $DBNAME
# --multiaddrs $MULTIADDR
# -q $GENESIS_TIME,$DEPOSIT_COUNT or -q $GENESIS_STATE_FILE
# optional
# -v $VALIDATOR_COUNT or -v $START_INDEX,$END_INDEX
# --dump true # dump state/blocks to disk
# --peer-id $PEER_ID_FILE

# examples

# create peer id json
./bin/create-peer-id p1.json

# no validators
./bin/lodestar interop -p minimal --db l2 -q 1567777777,16 --peer-id p1.json --multiaddrs /ip4/0.0.0.0/tcp/30607

# 16 validators
./bin/lodestar interop -p minimal --db l1 -q 1567777777,16 --multiaddrs /ip4/0.0.0.0/tcp/30606 -v 16
```

Working:
- chain quickstart
- block proposal
- networking

Not working:
- attestations

### CODE REVIEW

File: packages/lodestar/test/e2e/chain/factory/block/assembleBlock.test.ts [hunk 1]
```diff
@@ -108,11 +112,11 @@ describe('produce block', function () {
...
     blockProposingService.getRpcClient().validator.produceBlock.callsFake(async(slot, randao) => {
       // @ts-ignore
-      return await assembleBlock(config, dbStub, opPoolStub, eth1Stub, slot, randao);
+      return await assembleBlock(config, chainStub, dbStub, opPoolStub, eth1Stub, slot, randao);
     });
     const block = await blockProposingService.createAndPublishBlock(1, state.fork);
-
-    expect(() => stateTransition(config, state, block, false)).to.not.throw();
+    stateTransition(config, state, block, false)
+    //expect(() => stateTransition(config, state, block, false)).to.not.throw();
```
Comment (mpetrunic) [inline packages/lodestar/test/e2e/chain/factory/block/assembleBlock.test.ts]: ?

File: packages/lodestar/test/e2e/network/network.test.ts [hunk 2]
```diff
@@ -62,6 +62,8 @@ describe("[network] network", () => {
       new Promise((resolve) => netA.on("peer:disconnect", resolve)),
       new Promise((resolve) => netB.on("peer:disconnect", resolve)),
     ]);
+    await new Promise(resolve => setTimeout(resolve, 100));
```
Comment (mpetrunic) [inline packages/lodestar/test/e2e/network/network.test.ts]: lol. fyi there is sleep function somewhere so it's clearer what the intention is

File: docker-compose.yml [hunk 3]
```diff
@@ -0,0 +1,17 @@
+version: '2'
+services:
+    prometheus:
+        image: prom/prometheus:v2.9.2
```
Comment (mpetrunic) [inline docker-compose.yml:4]: add
```
   depends_on:
      - lodestar
```

File: docker/Dockerfile [hunk 4]
```diff
@@ -1,37 +1,16 @@
...
-COPY packages/lodestar/package.json ./packages/lodestar/package.json
-COPY yarn.lock .
+WORKDIR /root/lodestar
 
-RUN yarn install --frozen-lockfile && node_modules/.bin/lerna bootstrap && yarn cache clean
+RUN yarn
 
-COPY . .
+#ENTRYPOINT ["./packages/lodestar/bin/lodestar"]
+CMD ./packages/lodestar/bin/lodestar interop -p minimal --db l1 -q 1568584956,8 --multiaddrs /ip4/127.0.0.1/tcp/30607 -r -v 8
```
Comment (mpetrunic) [inline docker/Dockerfile:12]: Can we keep old file as Dockerfile.production. I will fix those images at some point again

File: packages/bls/package.json [hunk 5]
```diff
@@ -52,7 +52,7 @@
     "@babel/register": "^7.0.0",
     "@babel/runtime": "^7.4.4",
     "@chainsafe/benchmark-utils": "^0.1.0",
-    "@chainsafe/eth2.0-spec-test-util": "^0.2.3",
+    "@chainsafe/eth2.0-spec-test-util": "0.4.1",
     "@types/assert": "^1.4.2",
     "@types/chai": "^4.1.7",
     "@types/mocha": "^5.2.5",
```
Comment (mpetrunic) [inline packages/bls/package.json:58]: mocha types should be inline with mocha version

File: packages/lodestar/bin/create-peer-id [hunk 6]
```diff
@@ -0,0 +1,3 @@
+#!/usr/bin/env node
+
+require('../lib/interop/createPeerId');
```
Comment (mpetrunic) [inline packages/lodestar/bin/create-peer-id:3]: why isn't this a part of cli? :)

File: packages/lodestar/package.json [hunk 7]
```diff
@@ -92,6 +96,7 @@
     "pull-pushable": "^2.2.0",
     "pull-stream": "^3.6.13",
     "qs": "^6.7.0",
+    "queue": "^6.0.1",
```
Comment (mpetrunic) [inline packages/lodestar/package.json]: I dont think we need both queue and fast priority queue

File: packages/lodestar/peerId.json [hunk 8]
```diff
@@ -0,0 +1,5 @@
+{
+  "id": "16Uiu2HAmE129ZRUtTNtXHQTRgwnjMqKkgJWZb5LSqcx5br8YqxA8",
+  "privKey": "CAISIP5MQv0pH0dDqkDhyO88sUtpdfDn3MMVkBsS++GJlCnY",
+  "pubKey": "CAISIQMT/VX5iUjCV2Eia3iAUtO2lqiMaBFldHfnwrvLCgQ13Q=="
```
Comment (mpetrunic) [inline packages/lodestar/peerId.json]: Is there a reason for this to be commited?

File: packages/lodestar/src/chain/chain.ts [hunk 9]
```diff
@@ -84,35 +101,156 @@ export class BeaconChain extends (EventEmitter as { new(): ChainEventEmitter })
...
+
+    if(block.slot <= this.latestState.slot) {
+      this.logger.warn(`Block ${blockHash.toString("hex")} is in past. Probably fork choice/double propose/processed block. Ignored for now.`);
+      return;
+    }
+
+    if(block.slot > this.latestState.slot) {
+      //either block came too early or we are suppose to skip some slots
+      const latestBlock = await this.db.block.getChainHead();
+      if(!block.parentRoot.equals(signingRoot(latestBlock, this.config.types.BeaconBlock))){
```
Comment (mpetrunic) [inline packages/lodestar/src/chain/chain.ts:156]: duplicate code, same is in processBlock

File: packages/lodestar/src/chain/chain.ts [hunk 10]
```diff
@@ -84,35 +101,156 @@ export class BeaconChain extends (EventEmitter as { new(): ChainEventEmitter })
...
+      this.dumpState(block, pre, post);
+    }
+
+    this.metrics.currentSlot.inc(1);
 
     // forward processed block for additional processing
     this.emit('processedBlock', block);
+  };
+
+  public dumpState(block: BeaconBlock, pre: BeaconState, post: BeaconState): void {
```
Comment (mpetrunic) [inline packages/lodestar/src/chain/chain.ts]: I think we should move this. Maybe do it on block processed event or something

File: packages/lodestar/src/interop/dump.ts [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+  output: string,
+): void {
+  const tree = ProgressiveMerkleTree.empty(DEPOSIT_CONTRACT_TREE_DEPTH);
+  const state = quickStartState(config, tree, genesisTime, validatorCount);
+  const BeaconState = config.types.BeaconState;
+  fs.writeFileSync(output, serialize(state, BeaconState));
+}
+
+// eslint-disable-next-line
+import yargs from "yargs";
```
Comment (mpetrunic) [inline packages/lodestar/src/interop/dump.ts]: why isn't import on top of the file ?

Comment (mpetrunic) [reply packages/lodestar/src/interop/dump.ts]: why not use current cli?

File: packages/lodestar/src/interop/keypairs.ts [hunk 12]
```diff
@@ -14,13 +14,17 @@ interface IKeypair {
...
-      pubkey,
-    };
+    return interopKeypair(i);
   });
 }
+
+export function interopKeypair(index: number): IKeypair {
+  const privkey = bytesToBN(hash(intToBytes(index, 32)))
+    .mod(CURVE_ORDER)
+    .toArrayLike(Buffer, 'be', 32);
```
Comment (mpetrunic) [inline packages/lodestar/src/interop/keypairs.ts]: We should move this to bls(Private kes) so we don't have curve order constant in multiple places

File: packages/lodestar/src/logger/winston.ts [hunk 13]
```diff
@@ -17,9 +18,9 @@ export class WinstonLogger implements ILogger {
       ...options,
     };
     this.winston = createLogger({
-      level: LogLevel[LogLevel.debug], // log level switching handled in `createLogEntry`
+      level: LogLevel[LogLevel.verbose], // log level switching handled in `createLogEntry`
```
Comment (mpetrunic) [inline packages/lodestar/src/logger/winston.ts:21]: why double conversionj? field should be string anyways

File: packages/lodestar/src/metrics/options.ts [hunk 14]
```diff
@@ -11,8 +11,31 @@ export interface IMetricsOptions {
...
+      cli: {
+        flag: "metricsPort",
+      },
+    },
+    {
+      name: "boolean",
+      type: "bool",
+      configurable: true,
+      cli: {
+        flag: "metricsEnabled"
```
Comment (mpetrunic) [inline packages/lodestar/src/metrics/options.ts:32]: think flag is suppose to have `--` prefix

File: packages/lodestar/src/metrics/server/http.ts [hunk 15]
```diff
@@ -27,9 +30,14 @@ export class HttpMetricsServer implements IMetricsServer {
     }
   }
   public async start(): Promise<void> {
-    await promisify(this.http.listen.bind(this.http))(this.opts.serverPort);
+    if (this.opts.enabled) {
+      this.logger.info("Starting prometheus...");
```
Comment (mpetrunic) [inline packages/lodestar/src/metrics/server/http.ts]: and more information like on which host and port

**Review by mpetrunic — COMMENTED** [review]
¯\_(ツ)_/¯

File: packages/lodestar/src/logger/winston.ts (hunk 13 above)
Comment (wemeetagain) [reply packages/lodestar/src/logger/winston.ts:21]: enum gets converted to something like this at runtime:
```
{
  1: "debug",
  2: "verbose:,
  "debug: 1,
  "verbose: 2,
}
```
So the above is a roundabout way of doing:
```
  level: "verbose",
```
I think it shouldn't actually matter what `level` gets set to here, since according to the comment, the level gets written in `createLogEntry`

File: packages/lodestar/src/chain/chain.ts (hunk 9 above)
Comment (wemeetagain) [reply packages/lodestar/src/chain/chain.ts:156]: when we receive a block frp, the future, we want to see if this is the "next" block or not
if it isn't, we stick it back in the queue

File: docker-compose.yml (hunk 3 above)
Comment (wemeetagain) [reply docker-compose.yml]:
```suggestion
        image: prom/prometheus:v2.9.2
        depends_on:
            - lodestar
```

File: packages/bls/package.json (hunk 5 above)
Comment (wemeetagain) [reply packages/bls/package.json:58]: there is no published @types/mocha@6.x.x

File: packages/lodestar/bin/create-peer-id (hunk 6 above)
Comment (wemeetagain) [reply packages/lodestar/bin/create-peer-id:3]: shrug, we need an overhaul of our cli in general, leaving as is for now.

File: packages/lodestar/peerId.json (hunk 8 above)
Comment (wemeetagain) [reply packages/lodestar/peerId.json]: removing, it was included for interop

File: packages/lodestar/src/interop/dump.ts (hunk 11 above)
Comment (wemeetagain) [reply packages/lodestar/src/interop/dump.ts]: moving to top of file. leaving big changes for cli refactor.

File: packages/lodestar/src/metrics/options.ts (hunk 14 above)
Comment (wemeetagain) [reply packages/lodestar/src/metrics/options.ts:32]: gets added in src/cli/util.ts

File: packages/lodestar/src/chain/chain.ts (hunk 10 above)
Comment (wemeetagain) [reply packages/lodestar/src/chain/chain.ts]: not sure what to do, leaving for now

File: packages/lodestar/src/interop/keypairs.ts (hunk 12 above)
Comment (wemeetagain) [reply packages/lodestar/src/interop/keypairs.ts]: not sure what to do, leaving for now

**Comment by wemeetagain** [comment]
ready for a final review, all tests passing~

File: packages/lodestar/src/chain/stateTransition/util/attestation.ts [hunk 16]
```diff
@@ -117,10 +116,8 @@ export function getAttestingIndices(
   bits: BitList
 ): ValidatorIndex[] {
   const committee = getCrosslinkCommittee(config, state, data.target.epoch, data.crosslink.shard);
-  assert(bits.bitLength === committee.length);
-
   // Find the participating attesters in the committee
-  return committee.filter((_, i) => bits.getBit(i)).sort((a, b) => a - b);
+  return committee.filter((_, i) => {return bits.getBit(i);}).sort((a, b) => a - b);
```
Comment (wemeetagain) [inline packages/lodestar/src/chain/stateTransition/util/attestation.ts]: can this line stay the same?

File: packages/lodestar/src/chain/stateTransition/util/attestation.ts [hunk 17]
```diff
@@ -117,10 +116,8 @@ export function getAttestingIndices(
   bits: BitList
 ): ValidatorIndex[] {
   const committee = getCrosslinkCommittee(config, state, data.target.epoch, data.crosslink.shard);
-  assert(bits.bitLength === committee.length);
```
Comment (wemeetagain) [inline packages/lodestar/src/chain/stateTransition/util/attestation.ts:120]: why is this removed?

File: packages/lodestar/src/chain/stateTransition/util/attestation.ts (hunk 16 above)
Comment (wemeetagain) [inline packages/lodestar/src/chain/stateTransition/util/attestation.ts]: why do we need the return here?

File: packages/lodestar/src/logger/winston.ts [hunk 18]
```diff
@@ -28,9 +29,27 @@ export class WinstonLogger implements ILogger {
...
+              const infoString = (info.module || info.namespace || "");
+              const infoPad = paddingBetweenInfo - infoString.length;
+
+              // if (info.message.length + paddingPreMsg > screenSize) {
+              //   const max = screenSize - paddingPreMsg;
+              //   const p2 = info.message.length - max;
+              //   const msgs = [info.message.substring(0, max), info.message.substring(max)];
+              // eslint-disable-next-line max-len
+              //   return (`${info.timestamp}  [${infoString.toUpperCase()}] ${info.level.padStart(infoPad)}: ${msgs[0]}\n${msgs[1].padStart(paddingPreMsg + 2)}`)
+              // } else {
```
Comment (wemeetagain) [inline packages/lodestar/src/logger/winston.ts]: what are these comments for? still needed?

File: packages/lodestar/src/chain/stateTransition/util/attestation.ts (hunk 17 above)
Comment (mpetrunic) [reply packages/lodestar/src/chain/stateTransition/util/attestation.ts:120]: It's not in spec:
https://github.com/ethereum/eth2.0-specs/blob/v0.8.3/specs/core/0_beacon-chain.md#get_attesting_indices

File: packages/lodestar/src/chain/stateTransition/util/attestation.ts (hunk 16 above)
Comment (mpetrunic) [reply packages/lodestar/src/chain/stateTransition/util/attestation.ts]: we don't, I was logging something there and didn't return to original

File: packages/lodestar/src/logger/winston.ts (hunk 18 above)
Comment (mpetrunic) [reply packages/lodestar/src/logger/winston.ts]: I thinks thats @GregTheGreek log padding?

Comment (GregTheGreek) [reply packages/lodestar/src/logger/winston.ts]: Hahaha , yeah that's mine. We can remove it.

File: packages/lodestar/src/chain/stateTransition/util/attestation.ts (hunk 17 above)
Comment (wemeetagain) [reply packages/lodestar/src/chain/stateTransition/util/attestation.ts:120]: ok, looks like we're missing the length check here:
 in https://github.com/ethereum/eth2.0-specs/blob/v0.8.3/specs/core/0_beacon-chain.md#attestations

**Review by wemeetagain — APPROVED** [review]

**Review by ec2 — APPROVED** [review]
